Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
+    let diagnostics: [FlattenDiagnostic]
+
+    /// The content of the source line.
+    ///
+    /// This string contains the actual code from the source file at the specified line number, serving
+    /// as a reference point for diagnostics and annotations.
+    let sourceString: String
+
+    /// The line number in the source file corresponding to `sourceString`.
+    ///
+    /// This value facilitates easy navigation to and within the source file, providing a clear reference
+    /// to where the source line is located within its broader context.
+    let sourceLineNumber: Int
+
+    /// Additional text appended to the source line for extra context or information.
+    ///
+    /// This text can include annotations, explanations, or excerpts from relevant macro expansions,
+    /// aimed at providing further insight or clarification.
+    let suffixText: String
+
+    /// Determines if the source line lacks any associated diagnostics or annotations.
+    ///
+    /// This property returns `true` when there are no diagnostics related to the line and when the
+    /// `suffixText` is empty, indicating that the line is clean or unannotated. It offers a quick
+    /// way to assess the line's status in relation to issues or annotations.
+    var isFreeOfAnnotations: Bool {
+      return diagnostics.isEmpty && suffixText.isEmpty
     }
-    return annotatedSource
+  }
+}
+
+// Extension providing all necessary APIs to maintain backward compatibility. In the future,
+// upon committing to the new API of `DiagnosticsFormatter` and diagnostic decorators,
+// all members will be marked as deprecated.
+extension DiagnosticsFormatter {
+  public var contextSize: Int {
+    switch contextRange {
+    case .limited(let size):
+      return size
+    case .full:
+      return -1
+    }
+  }
+
+  @available(*, deprecated, message: "Store the `colorize` property passed to the initializer instead")
+  public var colorize: Bool {
+    diagnosticDecorator is ANSIDiagnosticDecorator
+  }
+
+  public init(contextSize: Int = 2, colorize: Bool = false) {
+    self.init(contextRange: .limited(contextSize), diagnosticDecorator: colorize ? .ANSI : .basic)
   }
 
-  /// Print given diagnostics for a given syntax tree on the command line
   public func annotatedSource(
     tree: some SyntaxProtocol,
     diags: [Diagnostic]
   ) -> String {
-    return annotatedSource(
-      tree: tree,
-      diags: diags,
-      indentString: "",
-      suffixTexts: [:]
+    annotatedSource(inSyntaxTree: tree, withDiagnostics: diags)
+  }
+
+  public static func annotatedSource(
+    tree: some SyntaxProtocol,
+    diags: [Diagnostic],
+    contextSize: Int = 2,
+    colorize: Bool = false
+  ) -> String {
+    Self.annotatedSource(
+      inSyntaxTree: tree,
+      withDiagnostics: diags,
+      limitedToContextRange: .limited(contextSize),
+      withDiagnosticDecorator: colorize ? .ANSI : .basic
     )
   }
 }
+
+extension Sequence where Element == Range<Int> {
```

Comment:
Since you’re refactoring this already, I think this can be shortened to 

```suggestion
extension Sequence<Range<Int>> {
```